### DISCUSSION TITLE
Changes to Matrix Blocks and CKEditor Blocks should be part of a draft

### DISCUSSION THREAD

**Discussion by maxstrebel** [craftcms/cms#14603] [General]
If I edit a page with matrix blocks or blocks in CK Editor I would assume all changes to be local/temporary before I saved. No matter if I open an edit dialog or not. Unlike Entry Relations, which are relations.

What actually happens is: The change in the slideout menu is on the live site right away – not part of my draft.

https://github.com/craftcms/cms/assets/910594/c5505894-e9d5-421e-a2be-e067fd0c958c

From an authors point of view I would assume that all changes to this page are part of my draft. If I want to propose a revision of that page to my team, I want to be able to show it with changes to my blocks, too (without them being on the live site). It also adds a lot of risk, as in the past clients would let their work students do these kinds of jobs (without the need of technical schooling) – which could now accidentally edit the live page.

The current assumption (https://github.com/craftcms/cms/discussions/14286) is, that the card view is an entry relation due to its definition. I would like to challenge this implementation. I agree that entries should always behave in the same and reliable way. An entry relation should open in the slideout and the change be applied to it. As a possible (theoretical) solution one could propose a differentiation between blocks (used in Matrix and CK Editor) and Entry Relations.

Thus I'd be very happy if changes to blocks are part of the provisional drafts and drafts.

**Comment by MattWilcox** [comment]
Just to add in here so it's in one place:

- My view is the flow in the video feels like a bug, not an intentional choice
- I think this behaviour will cause a lot of support requests from clients confused why content is being effected outside of a draft
- IMO, a Draft should create a new context, and changes made inside one should never leak out
- "Relational" stuff should be through Fields, and not through Entry Types. i.e., yes, if this was an Entry field's content being edited then the changes reflecting through make sense. But not for an Entry Type Instance. I do not exoect Entry Type Instances to behave like this.

**Comment by thupsi** [comment]
Some of these concerns have been expressed [in this discussion](https://github.com/craftcms/cms/discussions/14286), where I did not comment, because I thought it would be too much of an ask to negotiate changes of this nature so close to the Craft 5 launch. 

Nevertheless, I agree with all of the above and don't have much to add myself. I share the belief that the current behaviour can easily lead to unintended changes going live, and I don't expect authors to think of nested cards as related entries, especially in "content builder" scenarios. 

The only case where I would expect changes in nested entries to go live immediatelly would be in the "element index" mode, and even then, it could depend on the kind of relations, entries etc.

Would it be possible to have an option, per matrix field, to either "**save changes immediately**" or "**participate in owner drafts"**?

**Comment by MoritzLost** [comment]
@brandonkelly Can you comment on if/when those changes will be made to the behaviour for element cards in CKEditor and especially Matrix Blocks? We wanted to use the element cards display mode for our page builder Matrix field, but this issue makes this mode completely unusable for this use-case. Your comment https://github.com/craftcms/cms/discussions/14286#discussioncomment-8374856 sounds rather like this is the intended behaviour and there are no plans to change it?

**Reply by thupsi** [reply]
Hello @MoritzLost! 

Brandon is of course the one to give a definite answer, but in the meantime I wanted to give you an update from another discussion: https://github.com/craftcms/cms/discussions/14727#discussioncomment-8949790. TL;DR it seems we will get an improved behaviour, in line with what your editors would expect from a page builder. Just keep in mind that the roadmap item mentioned in my comment wasn't ready in time for `5.1` and is now slated for `5.x`.

**Reply by MoritzLost** [reply]
@thupsi Thanks! I say your comment there since it's linked on the roadmap, but that was back in March before 5.1, and `5.x` is scary to me 👀  I've seen some feature requests end up on the backburner for a long time, so I'm hoping for an update from Brandon if we can expect changes in the near future.

**Reply by maxstrebel** [reply]
Here's the end of the discusssion on discord: https://discord.com/channels/456442477667418113/456442477667418115/1218258609360601158. I am also waiting for this and will not upgrade to 5 before it is done ;)

**Reply by brandonkelly** [reply]
@maxstrebel Nothing changes if you keep using the inline-editable blocks view mode.

**Reply by maxstrebel** [reply]
Sure. I assumed that this change would come 5.X and really like CK Blocks and Cards 😉

**Comment by brandonkelly** [comment]
I’ve been toying with this, but… it’s complicated. There are a couple aspects of how things work that I don’t want to veer away from:

- We want to retain the ability to save nested content directly. Half of the reason we added the card view was so you can go in and make edits directly to nested content, without needing to save the entire top-level entry. That’s a major performance/AX win, especially on projects with more complex content models and/or a large number of sites.
- We don’t want the slideout to drastically differ from the full-page edit form; otherwise the entire concept of element editor slideouts starts to get a little muddy. They are UI-optimized versions of their full-page counterparts; nothing more or less.

That said, purely from a workflow perspective, I can see the benefit of being able to make edits across multiple nested entries, and having them all published at once.

The original plan discussed in Discord was:

1. Optimize the slideouts for closing without fully saving when done editing (“Done” as the primary action instead of “Save”).
2. Show provisional changes in the Matrix/CKEditor field’s cards, as an indication of nested pending changes.
3. Publish provisional changes when the parent entry is saved.

However:

- Changing the primary action within the slideout chips away at the element slideout concept, and would differ from the full-page edit form for nested entries.
- Button inconsistency aside, we think <kbd>Command</kbd>/<kbd>Ctrl</kbd> + <kbd>S</kbd> should definitely continue to directly save the nested content, but it would be very awkward if there was a discrepancy between that shortcut and the primary button.
- Automatically applying provisional drafts when saving the parent entry will likely have unintended consequences when authors didn’t expect their provisional drafts to go live yet.
- It would almost certainly not be expected behavior for nested content more than one level deep, whose content wouldn’t have any representation on the top-level page. So we should not do that. But now we’re back at square one for those entries.

So what do we do?

As a starting point, it feels like showing provisional changes within element cards/chips—across the entire control panel—is an obvious improvement. So that part is already done (pending a11y review): #14975.

For all slideouts capable of autosaving, we are also considering changing the “Close” button to “Done”. Coupled with the chips/cards showing provisional content, we think that will help communicate that it’s perfectly safe to close the slideout without losing your changes.

That leaves the problem of not having an easy way to see and publish all your nested changes at once, which we’re still thinking about.

**Reply by maxstrebel** [reply]
Thank you for the detailed answer. I understand the performance improvements and know why you want to address them. But I am afraid that this solution is very complex – technically but also to teach to clients.

I hope you consider to distinguish between matrix/ck editor blocks and entry relations at some point (as it used to be with Matrix).

In my experience, UI/UX wise matrix was never broken. My clients gave a lot of praise for its usability. The Cards and CK blocks (including the slideout interface) make it even better, because you can make stuff less messy. So I'd love to use them 🙂

**Reply by thupsi** [reply]
As a sidenote, I would like to highlight the fact that the drafts problem applies almost exclusively when cards are used for page builders or the like. A popular use case for sure.

There _are though_ cases where matrix fields are meant to contain autonomous entries and this is a powerfull new feature in Craft 5. We need to aknowledge this, to prevent fixing one thing and breaking another.

Consider this:

I have a single page called "Our Team". In this page I want to present 20 team members. Each team member must also have a dedicated page with meta, some text etc.

Prior to Craft 5, I would implement this by having a single for the "Our Team" page, and a structure for the team members. This would present the problem of having content with parent and child relationship in two different places and making my entry index more complicated. If instead I chose to have the team members in a matrix inside the "Our Team" page, it could get unwieldy, I would have to manually add routes for the member pages, I wouldn't be able to edit each member in it's own draft, and accessing them from other templates would be somewhat cumbersome. 

In Craft 5, I can have the team members in a matrix field and have all the benefits of separate entries without having to jump through hoops, plus also having the bonus of collocation and easy overview within the "Our Team" page, reflecting the actual mental model of the editors.

In short, the newfound ability of Matrix blocks to function as _nested sections_ solves part of what the Unified Content View was/will try to solve. This complicates things, because we can't say anymore that Matrix blocks _should always_ belong to the owner entry and be attached to its drafts. Rather, it depends.

**Reply by thupsi** [reply]
Another sidenote. I'm curious if the upcoming "Content Releases" feature could be leveraged to solve this conundrum.

**Reply by MoritzLost** [reply]
@brandonkelly Thinking through the challenges you outlined, I think the behaviour for nested entries should be an option on Matrix fields. I definitely agree that there shouldn't be two primary actions for pop-outs, this is too confusing.

I see two distinct use-cases, with different expectations for nested entries:

1. The nested entries _primarily belong to the entry_. For example, a page builder.
2. The nested entries are _independent entities_. @thupsi gave a good example with the team index page and individual team members.

For the first use-case, I expect the nested entries to have a _shared lifecycle_ with the parent entry. That is, changes stay provisional until I'm saving the parent entry, at which point all the provisional changes to the nested entries are saved as well. For the second use-case, I expect the nested entries to have their _own independent lifecycle_. This would be the current behaviour, where saving an entry from a pop-out will immediately update the canonical version.

Then there are some cases where this is not clear-cut, where nested entries kind of belong to the parent, but may also have their own frontend route or be used in other contexts.

I don't think there's a single solution that will work for all use-cases. In my opinion, those two modes of operation should be an option on matrix fields, so I as a developer can decide which mode of operation I want for every matrix field. I imagine a configuration on matrix fields with two options:

- **Nested entries share their owner's life-cycle.** Changes in a pop-out are saved as provisional drafts, and applied when the parent entry is saved.
- **Nested entries have their own life-cycle.** Changes to nested entries are applied immediately upon saving.

Wording can be improved, but you get the idea. This would also solve the issue with multiple levels of nested entries, because I can set the behaviour I want for every Matrix field at every level.

**Reply by thupsi** [reply]
I was thinking something along those lines also, see my [comment above](https://github.com/craftcms/cms/discussions/14603#discussioncomment-8805257)

> Would it be possible to have an option, per matrix field, to either "save changes immediately" or "participate in owner drafts"?

Now that I think of it some more, and expanding on @MoritzLost 's thinking, maybe in the case where the Matrix field is set to behave as a classic matrix block, you could remove the "Save" button completely in slideouts and just leave an "OK" button that will create a draft to be published along with the parent entry's save. 

That leaves the problem of full screen editing of those "block" entries, where it would be weird to have an OK button. I can think of two options:

1. Disable full screen editing for "block" entries.
2. Have a sort of notification on save (like the one for provisional drafts) that goes something like "This is nested content, your changes will apply when you save the top level parent".

Edit: Actually if you want to keep the save button in slideouts, you could have such a notification as described above.

**Reply by MoritzLost** [reply]
@brandonkelly What do you think about adding an option as I suggested in my comment above?

I understand this may not be a priority, though for some use-cases this is a vital issue. Could you let us know if this is still on the roadmap and if/when we could expect this issue to be addressed in some way? We're planning our Craft 5 updates, and we would love to be able to utilize the new element cards, but this lifecycle issues makes this impossible for most use-cases.

**Comment by MattWilcox** [comment]
I think the problem is the assumed context. It feels like Brandon and team P&T are thinking as though the most nested element is always the contextual reference - whereas when I'm editing a page *the page is the important context* and everything else is subject to that. i.e., if the page is a draft *everything in it is also a draft*. I think in terms of *the page* because I am in the CMS's page editing context.

The implementation of Cards really muddies that if the content of the card isn't isolated to the same context.

Tough one.

In my head, the behaviour Brandon describes is something I'd expect in a *relational field* like an Entry field. But not as part of an Entry Type. Which is not a field. It's its own thing.

**Reply by thupsi** [reply]
An entry though is not always a page. It could be a taxonomy group for example. It could be a global settings group. It could be lots of things where it's not the most important context. I'm playing the devil's advocate here, but really I'm trying to poke at this from all angles.

**Reply by MattWilcox** [reply]
That's my point though about "the context". In my head, you set the context of what you are doing from whichever section you're editing in - and the fact that Entries (the New Thing and not The Old Entry Concept) don't behave differently because they don't take into account the context within which they're used is "the issue".

I'm sort-of saying that the idea that Entries (The New Thing not The Old Thing) are applicable in many places as *the same instance* instead of *their own instance* is a fundamental problem.

**Reply by thupsi** [reply]
I think, ultimately, the correct "context", can only be known by the developer, so, the nested entry/block's behaviour [should be configurable](https://github.com/craftcms/cms/discussions/14603#discussioncomment-9544850).

**Reply by maxstrebel** [reply]
Yes, I completely agree. I like the entrification and that they all share the same behavior. And that this makes them pretty exchangeable. 

I still have the gut feeling that making them all entries under the hood but distinguish between an entry (as in page or entry) and a block (as a block in CK or in a Matrix) would be a better choice – instead of making it configurable, what might make projects even more complex to set up and to get into. This would give it an option to move stuff between sections, but would solve a lot of the UX/AX/DX overhead because the slideout could show the info if it is a relation or a block. You can rely that a related entry can (and should) easily be saved right away and that a changed block will be part of the entry/draft/release. I know this poses challenges in performance / database / versioning.

**Comment by maxstrebel** [comment]
Hey Brandon, this ticket had quite some comments on the philosophy of how things should work. I took some time to test Craft 5 with some smaller clients.

So besides all technological topics, Craft CMS is a content management system. In the current state, I have to add a warning text to the CK Editor fields, that authors cannot make any changes to it when they do a draft because (button, image, quote, ...) changes will be live right away. Same is for Cards (although I can just not use it).

Documentation at my clients now is as follows: Ideally they should use drafts for new versions of the content (event page for next year, new education offers, etc.), because they keep the revision history, slug (and more). But if there's changes to a CK Editor field (with blocks), they need to either duplicate the block, or save a new entry and switch them.

It keeps getting harder to tell the benefits, because it is really confusing and feels wrong. My bigger clients want their interns to edit the content. And they understandably start to ask if we can get back to something easier.

Will changes to CK be part of a draft at some point? Or is this something that will stay this way for a longer time?

**Reply by thupsi** [reply]
I strongly agree that this is especially problematic for CKEditor fields (in addition to content builders as mentioned further above). I can hardly imagine a case where an editor would want versioning of CKEditor nested entries to be separate from the text content.

**Reply by thupsi** [reply]
On the other hand, when a matrix is used like a "nested section", like the screenshot below, then the nested entries should definetely have separate versioning:

![image](https://github.com/user-attachments/assets/1b99cdf1-f787-4ddb-9e85-9bad936f2734)

Sadly, nested entries lack a revision history. At least they have drafts. Is this maybe an oversight or a choice to keep DB sizes from getting out of hand due to a large number of revisions? (Sorry, that's a separate discussion, feel free to ignore this last part.)

**Reply by maxstrebel** [reply]
I agree on that. If the interface makes clear that it is a relation, then yes. My very personal stance is that the Craft CMS Team should draw the line between (Matrix) Blocks that belong to the content of the page (entry) and Relations. I know that everything should be an entry and that's fine for me. But I want it to make sense from an author point of view.

Your example is perfect: 
I'd use something as an apartment list as a relation element to an entry. Why should you use a Matrix for that? You would use Matrix, if you wanted to add further context to each apartment that should not be part of the apartment entry. Which is specific to the page you're working on. I see no reason why changes to Matrix should not be part of the draft*. Do you?

* besides technical considerations, but I'd rather prefer a solution that solves these without confusing users

**Reply by MoritzLost** [reply]
I completely agree with this. I fully understand the technical and UI challenges. But from an editor perspective, the current behaviour makes no sense at all. They don't have the developer-centric abstraction of 'this content block is an entry, it has a separate lifecycle'. Especially in a CKEditor context, for an editor the embedded entries are just pieces of content, with the expectation that they share the lifecycle of the parent entry. And they are correct, it should be possible to include changes to embedded entries in a draft without affecting the live entry (by default), so this can't just be fixed with training or field instructions.

With CKEditor it's the most urgent since there's no way around this. But it's also important for Matrix fields – the new cards view provides much better AX than inline blocks for a content builder. But because of this issue, we can't use it.

**Reply by thupsi** [reply]
> I'd use something as an apartment list as a relation element to an entry. Why should you use a Matrix for that?

I use a matrix here because I want the owner entry to represent the apartment index page, where index specific content, but also some global apartment options live. So the editor goes to "apartments" and everything is there, no need to go to a separate "apartments" structure or single. In other words, this partly solves the old "shallow content for devs, structured view for editors" problem that the (maybe upcoming?) Unified Content View will try to address.

**Reply by thupsi** [reply]
To clarify the above: The "apartments" matrix entries are not relations, they are the actual apartment entries, there is no other "apartments" section. The important thing here is that they clearly have independend lifecycles from the "owner". But this is _contextual_ and not a global pattern, thus I'm advocating for partitipation in owner drafts to be configurable.

**Reply by thupsi** [reply]
Or, to avoid complicating things, keep separate versioning only when the matrix is configured with "element index" view mode. Maybe that could be a good compromise.

**Reply by maxstrebel** [reply]
> > I'd use something as an apartment list as a relation element to an entry. Why should you use a Matrix for that?
> 
> I use a matrix here because I want the owner entry to represent the apartment index page, where index specific content, but also some global apartment options live. So the editor goes to "apartments" and everything is there, no need to go to a separate "apartments" structure or single. In other words, this partly solves the old "shallow content for devs, structured view for editors" problem that the (maybe upcoming?) Unified Content View will try to address.

Yeah, makes sense. Though this is a very specific use case. **Still also in this case**, I'd assume as the information is shown on this specific page, why would you NOT want it to be part of the draft?

With Drafts you could: Preview changes to the items with your client, save changes for later, prepare an entry for later publishing. 

Same with entry versioning. If you by accident made stupid changes, you cannot roll back. You'll have to go into each item and roll it back manually.

I do not want to hate. This is just stuff we had in Craft. And now it's gone. And neither my authors nor me want it to be gone.

**Reply by thupsi** [reply]
It's a little hidden (you have to open them in a new tab to see the complete UI), but you _can still use drafts_ in nested entries. In my example, when I make a change to an individual apartment I don't want that to be tied to the apartments index drafts or versions, that would feel wrong and unexpected.

The problem for some (most?) cases is that you would want the drafts of the owner to be in sync with the ones of nested entries.

**Reply by thupsi** [reply]
> Yeah, makes sense. Though this is a very specific use case. **Still also in this case**, I'd assume as the information is shown on this specific page, why would you NOT want it to be part of the draft?

The information is displayed in many places, not only in the apartments index.

**Reply by thupsi** [reply]
> It's a little hidden (you have to open them in a new tab to see the complete UI), but you _can still use drafts_ in nested entries.

<img width="1075" alt="image" src="https://github.com/user-attachments/assets/b408c994-a562-4568-a946-3ec520747742">

**Reply by maxstrebel** [reply]
So I'm NOT talking about nested entries :) they're completely fine as they are. I'm ONLY talking about Matrix Blocks and CK Editor blocks.

So apart from the obvious case of CK Editor, in 99% of the cases my clients use drafts to either review content before it goes live or the prepare a new revision of a page. Examples are:

- Event is over and the event page should show impressions from the event (completely new blocks)
- A page is a preview page but at day X a product becomes available. You want to prepare the page for day X and just apply the draft on day X
- You task an intern to do changes but a senior person has to proof before it goes live

All these are use cases that _every_ of my clients above a certain budget makes sure are present. And I have to make sure these workflows still work. In Craft 5 this means I can only use Inline Matrix Blocks and no CK Editor Blocks.

That said, CK Editor Versioning/Drafts is the main problem for me. As I cannot reasonably explain why changes are not part of the draft.

**Reply by thupsi** [reply]
As I said before, I agree with you. This may not be clear (my fault) because we got a little sidetracked with my particular example. In many cases (don't want to say most cases, because people can do all sorts of things with such a flexible system) _what you're asking should be the default_.

My "sidenotes" are fineprint:

- Nested entries and matrix blocks should continue to be able to be used interchangeably, it should be up to the dev to decide. You say that you don't want to lose existing abilities, but if we separate blocks from nested entries we shall lose _new abilities_.
- You _can_ still use your old way of doing things if you keep matrix fields in inline mode and don't use nested entries in CKEditor (keep using the old way of matrix fields and CKEditor as an inline block in them). I realise and agree this is not ideal, we want the best of both worlds.

**Comment by ryanleichty** [comment]
Sanity solves this by differentiating between _documents_ and _objects_. To quote someone else's explanation on the differences between the two:

> Object types cannot be represented as standalone documents in the data store. To define an object type to represent it as a document with an ID, revision, as well as created and updated timestamps, you should define with the document type. Apart from these additional fields, there’s no semantic difference between a document and an object.

Sanity's version of a matrix field is the _array_ schema type. To create a page builder like experience, you would use an _array_ of _object_ types. You can also use an array of document types, but documents are like entries in that they exist on their own with revisions, etc.

Not that I know how you would go about implementing something like this, but it seems like the need is for an _object_ type entry. I created [a discussion around this](https://github.com/craftcms/cms/discussions/15626) previously.

**Reply by ryanleichty** [reply]
We [added this issue](https://github.com/craftcms/cms/issues/15956) too since it feels like a bug from an authoring experience. Is there any kind of timeline in terms of finding a solve for this, or is it too early to tell?

**Comment by brandonkelly** [comment]
This is now resolved for Craft 5.5, via #16002 🎉

**Reply by brandonkelly** [reply]
Craft 5.5.0 and CKEditor 4.4.0 are out with those changes 🎉

**Reply by maxstrebel** [reply]
![nicholas-cage-fabulous](https://github.com/user-attachments/assets/742ce03b-b261-4451-aee5-eb10a239f2fe)

**Reply by MoritzLost** [reply]
@brandonkelly That's a huge improvement, thanks! Finally we can switch our content builder to cards view, so much better for the author experience 🎉

**Reply by dletsche** [reply]
Agreed! Huge improvement. Thank you so much!
